src/profile-logic/import/instruments/index.js

```diff
@@ -0,0 +1,910 @@
+    `run${selectedRun}`
+  );
+  return getOrThrow(corespaceRunDir.subdirectories, 'core');
+}
+
+// This function mainly extracts two entities: samples and arrays
+// Here arrays contains all the information about stack trace at a given timestamp.
+// Each samples has a field named 'backtraceID' which is an index into arrays
+// Iterating recursively into arrays by given backtraceID it extracts a backtraceStack for each sample
+async function getSamples(args: {
+  tree: TraceDirectoryTree,
+  addressToFrameMap: Map<number, InstrumentsFrameInfo>,
+  runNumber: number,
+}): Promise<Array<InstrumentsSample>> {
+  const { tree, addressToFrameMap, runNumber } = args;
+  const core = getCoreDirForRun(tree, runNumber);
+  const samples = await getRawSampleList(core);
+  const arrays = await getIntegerArrays(samples, core);
+  const backtraceIDtoStack = new Map<number, number[]>();
+
+  function appendRecursive(k: number, stack: Array<number>) {
+    const frame = addressToFrameMap.get(k);
+    if (frame) {
+      stack.push(k);
+      return;
+    }
+
+    const arrayIndex = k & 0xffffffff;
+    if (arrayIndex in arrays) {
+      for (const addr of arrays[arrayIndex]) {
+        appendRecursive(addr, stack);
+      }
+      return;
+    }
+
+    // Fallback: We are not able to find the address, so we will just
+    // display the address instead of frame information
+    const rawAddressFrame: InstrumentsFrameInfo = {
+      key: k,
+      name: `0x${zeroPad(k.toString(16), 16)}`,
+    };
+    addressToFrameMap.set(k, rawAddressFrame);
+    stack.push(k);
+  }
+
+  for (const sample of samples) {
+    getOrInsert(backtraceIDtoStack, sample.backtraceID, id => {
+      const stack: Array<number> = [];
+      appendRecursive(id, stack);
+      stack.reverse();
+      return stack;
+    });
+  }
+
+  for (const sample of samples) {
+    sample.backtraceStack = backtraceIDtoStack.get(sample.backtraceID);
+  }
+
+  return Promise.resolve(samples);
+}
+
+// This function reads the 'form.template' file which contains all the important information about
+// addressToFrameMap and Instruments' version
+async function readFormTemplateFile(tree) {
+  const formTemplate = tree.files.get('form.template');
+  const archive =
+    typeof formTemplate !== 'undefined'
+      ? readInstrumentsArchive(await readAsArrayBuffer(formTemplate))
+      : {};
+
+  const version = archive['com.apple.xray.owner.template.version'];
+  let selectedRunNumber = 1;
+  if ('com.apple.xray.owner.template' in archive) {
+    selectedRunNumber = archive['com.apple.xray.owner.template'].get(
+      '_selectedRunNumber'
+    );
+  }
+  let instrumentType = archive.$1;
+  if ('stubInfoByUUID' in archive) {
+    instrumentType = Array.from(archive.stubInfoByUUID.keys())[0];
+  }
+  const allRunData = archive['com.apple.xray.run.data'];
+
+  const runs: FormTemplateRunData[] = [];
+  for (const runNumber of allRunData.runNumbers) {
+    const runData = getOrThrow(allRunData.runData, runNumber);
```

Comment:
maybe it's a good idea to force-type `runData` or `allRunData.runNumbers` above, to assess what you expect this object to be. The reason is that `allRunData` is completely untyped because `archive` above is untyped.

Reply:
Added the type.